rtw_excel_report/models/product_spec.py

```diff
@@ -38,256 +35,425 @@ def generate_xlsx_report(self, workbook, data, lines):
 
-                        attr=""
-                        if attributes:
-                            for attribute in attributes :
-                                attr += attribute.attribute_id.name + ":" + attribute.product_attribute_value_id.name + "\n" 
-                            attr += size_detail
-                            sheet.write(index, 4, attr, wrap_format)
+                                
+                        ## 4 picture + 4 attributes
+                        if sol.product_id:
                             
-                            # take first 4 element of record
-                            attrs_has_img = [x for x in sl.product_id.product_template_attribute_value_ids if x.product_attribute_value_id.image][:4]
-
-                            if attrs_has_img:
-                                for index2,attr_img in enumerate(attrs_has_img) : 
-                                    image_attr_db= base64.b64decode(attr_img.product_attribute_value_id.image)
-                                    image = Image.open(io.BytesIO(image_attr_db))
-                                    path = os.path.join(image_attribute_product, f"{sl.id}_{index2}.png")
-                                    image.save(path, 'PNG')
-                                    img = PILImage.open(path)
-                                    img = img.convert('RGB')
-                                    img = img.resize((125, 125))
-                                    img_io = BytesIO()
-                                    img.save(img_io, 'JPEG')
-                                    img_io.seek(2)
-
-                                    # loop image with recipe
-                                    count_attr_img = index // 4  
-                                    row_attr_img = 31 + count_attr_img * 52
-                                    col_attr_img = (index % 4)*5
-
-                                    if index2 == 1:
-                                        col_attr_img += 2
-                                    elif index2 == 2:
-                                        row_attr_img +=8
-                                    elif index2 == 3:
-                                        row_attr_img +=8
-                                        col_attr_img += 2
+                            attr_child_count = 0
+                            attr_child_ids =[]
+                            for parent_attr in sol.product_id.product_template_attribute_value_ids.mapped('product_attribute_value_id'):
+                                if attr_child_count < 4:
+                                    if parent_attr.id not in attr_child_ids and parent_attr.image and all(attr not in sol.product_id.product_template_attribute_value_ids.mapped('product_attribute_value_id') for attr in parent_attr.child_attribute_ids.mapped('child_attribute_id')):
```

Comment:
cái if này nhìn gớm quá.
if parent_attr.id not in attr_child_ids
-> if parent_attr.id  in attr_child_ids:
continue
là được